pyne/source_sampling.pyx

```diff
@@ -372,10 +373,63 @@ cdef class Sampler:
             for ie_bounds in range(e_bounds_size):
                 e_bounds_proxy[ie_bounds] = <double> e_bounds[ie_bounds]
         self._inst = new cpp_source_sampling.Sampler(std_string(<char *> filename_bytes), std_string(<char *> src_tag_name_bytes), e_bounds_proxy, <bint> uniform)
-    
+
+    def _sampler_sampler_2(self, names, e_bounds, mode):
+        """Sampler(self, names, e_bounds, mode)
+        
+        Constuctor for overall Sampler
+        
+        Parameters
+        ----------
+        names : std::map<std::string, std::string>
+        
+        e_bounds : std::vector< double >
+        
+        Returns
+        -------
+        None
+        
+        ################################################################
+        
+        Parameters
+        ----------
+        e_bounds : std::vector< double >
+        
+        names : std::map<std::string, std::string>
+        
+        mode : int
+        
+        Returns
+        -------
+        None
+        
+        """
+        cdef cpp_vector[double] e_bounds_proxy
+        cdef int ie_bounds
+        cdef int e_bounds_size
+        cdef double * e_bounds_data
+        # Convert names
+        cdef cpp_map[std_string, std_string] cpp_names = cpp_map[std_string, std_string]()
+        for key, value in names.items():
+            cpp_names[key] = value
+
+        # e_bounds is a ('vector', 'float64', 0)
+        e_bounds_size = len(e_bounds)
+        if isinstance(e_bounds, np.ndarray) and (<np.ndarray> e_bounds).descr.type_num == np.NPY_FLOAT64:
+            e_bounds_data = <double *> np.PyArray_DATA(<np.ndarray> e_bounds)
+            e_bounds_proxy = cpp_vector[double](<size_t> e_bounds_size)
+            for ie_bounds in range(e_bounds_size):
```

Comment:
There are faster ways to copy array data into a vector than iterating over every element. In fact you have already done the hard part by getting the pointer in `e_bounds_data`. In fact I think that you can do:

```cython
e_bounds_proxy = cpp_vector[double](e_bounds_data, e_bounds_data + <size_t> e_bounds_size)
```

Reply:
There will be an error like this if I use the method you suggested:
```
Error compiling Cython file:
------------------------------------------------------------
...
        if isinstance(e_bounds, np.ndarray) and (<np.ndarray> e_bounds).descr.type_num == np.NPY_FLOAT64:
            e_bounds_data = <double *> np.PyArray_DATA(<np.ndarray> e_bounds)
            e_bounds_proxy = cpp_vector[double](<size_t> e_bounds_size)
            #for ie_bounds in range(e_bounds_size):
            #    e_bounds_proxy[ie_bounds] = e_bounds_data[ie_bounds]
            e_bounds_proxy = cpp_vector[double](e_bounds_data, e_bounds_data + <size_t> e_bounds_size)
                                                            ^
------------------------------------------------------------

/home/zxk/opt/pyne/pyne/source_sampling.pyx:370:61: Cannot assign type 'double *' to 'size_t'
```

Reply:
Sorry minor modification needed:

```cython
e_bounds_proxy = cpp_vector[double](<size_t> e_bounds_size)
memcpy(<void*> &e_bounds_proxy[0], e_bounds_data, sizeof(double) * e_bounds_size)
```

or similar. You'll also need to add the following to the top of the file.

```cython
from libc.string cimport memcpy
```

Reply:
Thanks for the suggestion. I made the `e_bounds` conversion part as a separate function to reduce duplications.